File: frontend/lib/traces.ts
import { API_BASE_URL } from "@/lib/auth";

export type TraceRecord = {
  id: number;
  conversation_id: number | null;
  user_id: number;
  intent: string;
  user_input: string;
  intent_data: Record<string, unknown>;
  retrieved_chunks: Array<Record<string, unknown>>;
  llm_input_summary: string | null;
  llm_output: string | null;
  tool_name: string | null;
  tool_args: Record<string, unknown>;
  approval_status: string;
  final_result: Record<string, unknown>;
  error_message: string | null;
  elapsed_ms: number;
  created_at: string;
};

async function parseTraceResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const fallback = response.status === 403 ? "当前账号没有追踪查看权限" : "追踪数据请求失败，请稍后重试";
    const body = await response.json().catch(() => null);
    throw new Error(typeof body?.detail === "string" ? body.detail : fallback);
  }

  return response.json() as Promise<T>;
}

function authHeaders(accessToken: string): HeadersInit {
  return {
    Authorization: `Bearer ${accessToken}`,
  };
}

export async function listTraces(accessToken: string): Promise<TraceRecord[]> {
  const response = await fetch(`${API_BASE_URL}/api/traces`, {
    headers: authHeaders(accessToken),
    cache: "no-store",
  });

  return parseTraceResponse<TraceRecord[]>(response);
}

export async function getTrace(accessToken: string, traceId: number): Promise<TraceRecord> {
  const response = await fetch(`${API_BASE_URL}/api/traces/${traceId}`, {
    headers: authHeaders(accessToken),
    cache: "no-store",
  });

  return parseTraceResponse<TraceRecord>(response);
}

File: backend/app/api/traces.py
import json

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_roles
from app.models.agent_trace import AgentTrace
from app.models.user import User
from app.schemas.trace import TraceRead
from app.services.trace_service import get_trace, list_traces


router = APIRouter()


def trace_to_read(trace: AgentTrace) -> TraceRead:
    return TraceRead(
        id=trace.id,
        conversation_id=trace.conversation_id,
        user_id=trace.user_id,
        intent=trace.intent,
        user_input=trace.user_input,
        intent_data=json.loads(trace.intent_json),
        retrieved_chunks=json.loads(trace.retrieved_chunks_json),
        llm_input_summary=trace.llm_input_summary,
        llm_output=trace.llm_output,
        tool_name=trace.tool_name,
        tool_args=json.loads(trace.tool_args_json),
        approval_status=trace.approval_status,
        final_result=json.loads(trace.final_result_json),
        error_message=trace.error_message,
        elapsed_ms=trace.elapsed_ms,
        created_at=trace.created_at,
    )


@router.get("", response_model=list[TraceRead])
def list_agent_traces(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(["admin"])),
) -> list[TraceRead]:
    return [trace_to_read(trace) for trace in list_traces(db)]


@router.get("/{trace_id}", response_model=TraceRead)
def get_agent_trace(
    trace_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(["admin"])),
) -> TraceRead:
    trace = get_trace(db, trace_id)
    if trace is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Trace not found",
        )
    return trace_to_read(trace)
